ParquetSharp.DataFrame/DataFrameExtensions.cs

```diff
@@ -8,58 +8,100 @@ namespace ParquetSharp
-                    using var logicalWriter = columnWriter.LogicalWriter();
-                    logicalWriter.Apply(new DataFrameWriter(dataFrameColumn, offset, batchSize));
-                }
-
-                offset += batchSize;
+                dataFrame.ToParquet(fileWriter, rowGroupSize);
             }
 
             fileWriter.Close();
         }
 
+        /// <summary>
+        /// Write a DataFrame in Parquet format
+        /// </summary>
+        /// <param name="dataFrame">The DataFrame to write</param>
+        /// <param name="path">Path to write to</param>
+        /// <param name="writerProperties">Optional writer properties that override the default properties</param>
+        /// <param name="logicalTypeOverrides">Mapping from column names to Parquet logical types,
+        /// overriding the default logical types. When writing decimal columns, a logical type must be provided
+        /// to specify the precision and scale to use.</param>
+        /// <param name="rowGroupSize">Maximum number of rows per row group</param>
+        public static void ToParquet(this DataFrame dataFrame, string path, WriterProperties? writerProperties = null,
+            IReadOnlyDictionary<string, LogicalType>? logicalTypeOverrides = null, int rowGroupSize = 1024 * 1024)
+        {
+            using ParquetFileWriter fileWriter = GetParquetFileWriter(path, writerProperties, logicalTypeOverrides, dataFrame);
+            dataFrame.ToParquet(fileWriter, rowGroupSize);
+
+            fileWriter.Close();
+        }
+
+        private static ParquetFileWriter GetParquetFileWriter(
+            string path, WriterProperties writerProperties,
+            IReadOnlyDictionary<string, LogicalType> logicalTypeOverrides, DataFrame firstDataFrame)
+        {
+            Column[] schemaColumns = firstDataFrame.Columns.Select(col => GetSchemaColumn(
+                col,
+                (logicalTypeOverrides != null && logicalTypeOverrides.TryGetValue(col.Name, out LogicalType logicalType))
+                ? logicalType : null)).ToArray();
+
+            return writerProperties == null
+                            ? new ParquetFileWriter(path, schemaColumns)
+                            : new ParquetFileWriter(path, schemaColumns, writerProperties);
+        }
+
         private static Column GetSchemaColumn(DataFrameColumn column, LogicalType? logicalTypeOverride)
         {
-            var dataType = column.DataType;
-            var nullable = column.NullCount > 0;
+            Type dataType = column.DataType;
             if (dataType == typeof(decimal) && logicalTypeOverride == null)
-            {
                 throw new ArgumentException($"Logical type override must be specified for decimal column '{column.Name}'");
-            }
 
-            if (nullable && dataType.IsValueType)
-            {
+            if (dataType.IsValueType && Nullable.GetUnderlyingType(dataType) != null)
```

Comment:
I think it's a bug to change a schema of parquet depending on data within a particular row.   If you are writing int? it should always be a nullable column even if there are no actual values.  Some readers may expect exact type.

Reply:
Yes I think it makes sense to always write value types as nullable, as DataFrame columns are always nullable. This is going to be required to support writing an `IEnumerable<DataFrame>` as otherwise there could be a scenario where the first file has no nulls but a subsequent file has nulls which would cause a crash, and we can't iterate over the `IEnumerable` multiple times to check for nulls unless we were more restrictive and only allowed writing `IReadOnlyList` for example. But making the type nullable depending on the null count was something I wasn't sure about when I wrote this and I think this scenario is a good example of why we should just always write value types as nullable.

This change looks wrong though. When you have a `DataFrameColumn` of `Int32` for example, `column.DataType` will be `System.Int32`, not `System.Nullable<System.Int32>`, even if the column contains null values. So I think you need to remove the `Nullable.GetUnderlyingType(dataType) != null` check as that will always be false. I think you'll find that this is causing some test failures, and you'll also need to update tests that expect non-nullable columns.

You'll also need to add the surrounding curly braces back in order to satisfy our code style check.